src/op_handlers/heap_read.rs

```diff
@@ -0,0 +1,31 @@
+use u256::U256;
+use zkevm_opcode_defs::MAX_OFFSET_TO_DEREF_LOW_U32;
+
+use crate::address_operands::address_operands_read;
+use crate::value::TaggedValue;
+use crate::{opcode::Opcode, state::VMState};
+
+pub fn _heap_read(vm: &mut VMState, opcode: &Opcode) {
+    let (src0, _) = address_operands_read(vm, opcode);
+    if src0.is_pointer {
+        panic!("Invalid operands for heap_read");
+    }
+
+    if src0.value > U256::from(MAX_OFFSET_TO_DEREF_LOW_U32) {
```

Comment:
I think its better to leave it as is, first because adding an extra function for 4-5 lines seems too much, even if they repeat, and second because this is going to eventually change when we add proper error handling, and then we will implement it in the best way possible